src/System.Text.RegularExpressions/src/System/Buffers/MemoryOrPinnedSpan.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Buffers
+{
+    internal unsafe readonly struct MemoryOrPinnedSpan<T>
+    {
+        private readonly ReadOnlyMemory<T> _memory;
+        private readonly char* _pinnedSpan;
+
+        public MemoryOrPinnedSpan(ReadOnlyMemory<T> memory)
+        {
+            _memory = memory;
+            Length = memory.Length;
+            _pinnedSpan = null;
+        }
+
+        public MemoryOrPinnedSpan(char* pinnedSpan, int length)
+        {
+            _pinnedSpan = pinnedSpan;
+            Length = length;
+        }
+
+        public static MemoryOrPinnedSpan<T> Empty { get; } = new MemoryOrPinnedSpan<T>();
+
+        public int Length { get; }
```

Comment:
The syntactic shortcut here is hurting clarity. It is important to see the state of fields like these together. This should be `int _length;` above; and the implementation should be `public int Length => _length`.

For example, we could have used the same syntactic shortcut in the implementation of the Span itself; but we intentionally do not: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.Portable.cs#L419